File: go/leetcode/03567.go
package main

import (
	"math"
	"slices"
	"sync"
)

func checkSubmatrix(grid [][]int, row int, col int, k int, ans [][]int) {
	vals := make([]int, 0, k*k)

	for dr := 0; dr < k; dr++ {
		vals = append(vals, grid[row+dr][col:col+k]...)
	}

	slices.Sort(vals)

	minDiff := math.MaxInt32
	foundUnique := false

	for i := 0; i < len(vals)-1; i++ {
		if vals[i] == vals[i+1] {
			continue
		}

		diff := vals[i+1] - vals[i]
		if diff < minDiff {
			minDiff = diff
			foundUnique = true
		}
	}

	if foundUnique {
		ans[row][col] = minDiff
	}
}

func minAbsDiff(grid [][]int, k int) [][]int {
	// Nice parallel implementation im proud of
	m := len(grid)
	n := len(grid[0])

	ans := make([][]int, m-k+1)

	// Initialize each inner slice
	for i := range ans {
		ans[i] = make([]int, n-k+1)
	}

	var wg sync.WaitGroup

	for row := range m {
		if row+k > m {
			break
		}
		wg.Add(1)
		go func(r int) {
			defer wg.Done()
			for c := range n {
				if c+k > n {
					break
				}
				checkSubmatrix(grid, r, c, k, ans)
			}
		}(row)
	}

	wg.Wait()

	return ans
}

File: python/LeetCode/03567.py
from typing import List

def minAbsDiff(self, grid: List[List[int]], k: int) -> List[List[int]]:
    m = len(grid)
    n = len(grid[0])

    ans = [([0] * (n - k + 1)) for _ in range(m - k + 1)]

    # Consider (row, col) to be the start (top left corner) of a submatrix
    for row in range(m):
        for col in range(n):
            unique_values = set()

            if row + k > m or col + k > n:
                break

            for r in range(k):
                for c in range(k):
                    unique_values.add(grid[row + r][col + c])

            unique_values = sorted(list(unique_values))

            if len(unique_values) < 2:
                break

            min_abs_diff = abs(unique_values[0] - unique_values[1])
            for v1, v2 in zip(unique_values, unique_values[1:]):
                min_abs_diff = min(min_abs_diff, abs(v1 - v2))

            ans[row][col] = min_abs_diff

    return ans